fiat-ldap-v2/src/main/java/com/netflix/spinnaker/fiat/roles/ldap/LdapUserRolesProviderV2.java

```diff
@@ -0,0 +1,212 @@
+import javax.naming.NamingException;
+import javax.naming.directory.Attribute;
+import javax.naming.directory.Attributes;
+import javax.naming.ldap.LdapName;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.tuple.Pair;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.dao.IncorrectResultSizeDataAccessException;
+import org.springframework.ldap.core.AttributesMapper;
+import org.springframework.ldap.core.DirContextOperations;
+import org.springframework.ldap.core.DistinguishedName;
+import org.springframework.ldap.support.LdapEncoder;
+import org.springframework.security.ldap.LdapUtils;
+import org.springframework.security.ldap.SpringSecurityLdapTemplate;
+import org.springframework.stereotype.Component;
+
+@Slf4j
+@Component
+@ConditionalOnProperty(value = "auth.group-membership.service", havingValue = "ldapv2")
+public class LdapUserRolesProviderV2 implements UserRolesProvider {
+
+  @Setter private SpringSecurityLdapTemplate ldapTemplate;
+
+  @Setter private LdapConfigV2.ConfigProps configProps;
+
+  public LdapUserRolesProviderV2(
+      @Autowired SpringSecurityLdapTemplate ldapTemplate,
+      @Autowired LdapConfigV2.ConfigProps configProps) {
+    this.ldapTemplate = ldapTemplate;
+    this.configProps = configProps;
+  }
+
+  @Override
+  public List<Role> loadRoles(ExternalUser user) {
+    String userId = user.getId();
+
+    log.debug("loadRoles for user " + userId);
+    if (StringUtils.isEmpty(configProps.getGroupSearchBase())) {
+      return new ArrayList<>();
+    }
+
+    String fullUserDn = getUserFullDn(userId);
+
+    if (fullUserDn == null) {
+      // Likely a service account
+      log.debug("fullUserDn is null for {}", userId);
+      return new ArrayList<>();
+    }
+
+    String[] params = new String[] {fullUserDn, userId};
+
+    if (log.isDebugEnabled()) {
+      log.debug(
+          new StringBuilder("Searching for groups using ")
+              .append("\ngroupSearchBase: ")
+              .append(configProps.getGroupSearchBase())
+              .append("\ngroupSearchFilter: ")
+              .append(configProps.getGroupSearchFilter())
+              .append("\nparams: ")
+              .append(StringUtils.join(params, " :: "))
+              .append("\ngroupRoleAttributes: ")
+              .append(configProps.getGroupRoleAttributes())
+              .toString());
+    }
+
+    // Copied from org.springframework.security.ldap.userdetails.DefaultLdapAuthoritiesPopulator.
+    Set<String> userRoles =
+        ldapTemplate.searchForSingleAttributeValues(
+            configProps.getGroupSearchBase(),
+            configProps.getGroupSearchFilter(),
+            params,
+            configProps.getGroupRoleAttributes());
+
+    log.debug("Got roles for user " + userId + ": " + userRoles);
```

Comment:
We wrapped line 84 in an if check, should we do the same here?

Reply:
I think yes... and also i wrap the same line in v1 module

Reply:
isn't the "if debug enabled" redundant?  e.g. I'd remove the if check from both as the logging service should handle that....